poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+
+> TODO: Describe how the permutation instance and the rate relate to each other and to the
+> desired security parameter `M`.
+
+`PoseidonSponge` has several internal properties:
+- An internal state of `T` field elements.
+- A mode variable that is either `absorbing` or `squeezing`.
+- A cache of field elements corresponding to the current mode.
+- A duplex padding function that is applied to input elements prior to incorporating them
+  into the sponge state.
+
+The only operations exposed by the sponge are:
+- `PoseidonSponge.Initialize`, which prepares a new sponge.
+- `PoseidonSponge.Absorb`, which takes a single field element into the sponge.
+- `PoseidonSponge.Squeeze`, which produces a single field element derived from all prior
+  `Absorb` and `Squeeze` operations on the sponge, as well as the initialization values.
+
+`PoseidonSponge.Initialize` must only be called once per sponge, and must be called first.
+After that, `PoseidonSponge.Absorb` and `PoseidonSponge.Squeeze` may be called repeatedly
+and in any order.
+
+The sponge has an internal state of width `T`, divided into two sections:
+- Field elements at indices `[0, RATE)` form the "rate" portion of the sponge.
+- Field elements at indices `[RATE, T)` (the last `C` field elements of the state) form
+  the capacity portion of the sponge. These elements are never directly affected by the
+  input field elements, and are never output during the squeezing phase.
+
+> Note: Both the duplex sponge construction from [[BDPV2011]], and the plain sponge
+> construction it is derived from [[BDPV2007]], use an internal state with the form
+> `rate || capacity`. Section 2.1 of the Poseidon paper matched this internal state form
+> when specifying the Poseidon sponge. However, Appendix I of the Poseidon paper (which
+> describes a particular use of the Poseidon sponge within Merkle tree hashing) used
+> `capacity || rate` instead, and at least one instance of Poseidon Merkle tree hashing
+> matching Appendix I was deployed into production before this inconsistency was noticed.
+>
+> TODO: Examine all open-source production instances of Poseidon to figure out how far
+> this inconsistency propagated, and what other variants might currently exist.
+
+[BDPV2007]: https://keccak.team/files/SpongeFunctions.pdf
+
+### Initialization
+
+`PoseidonSponge.Initialize` takes as input a `PoseidonPermutation`, a field element
+`domain_iv` that is used for domain separation, and a `duplex_padding` function. It
+proceeds as follows:
+
+1. Initialize the sponge state. The first element of the capacity portion is set to
+   `domain_iv`, while the remaining capacity elements, and all of the rate elements, are
+   initialized to zero.
+   ```
+   state = [0; T]
+   state[RATE] = domain_iv
+   ```
+
+2. Return the new sponge:
+   ```
+   PoseidonSponge(
+     PoseidonPermutation,
+     state,
+     mode = absorbing,
+     cache = [],
+     duplex_padding,
+   )
+   ```
```

Comment:
If there's a reasonably well-understood syntax we could use for structs (which I guess braces are), I'm happy to use that. Otherwise, I'll see if I can figure out some clear prose here.

Reply:
I don't think so, TLS uses their own made-up language for that, most specs just invent something. I'm a fan of the noise spec which just lists what's in the state (for example: http://www.noiseprotocol.org/noise.html#overview-of-handshake-state-machine)